connectors/src/connectors/google_drive/temporal/spreadsheets.ts

```diff
@@ -0,0 +1,380 @@
+  file: GoogleDriveObjectType | GoogleDriveFiles
+) {
+  return file.mimeType === "application/vnd.google-apps.spreadsheet";
+}
+
+function makeTableIdFromSheetId(spreadsheetId: string, sheetId: number) {
+  return `google-spreadsheet-${spreadsheetId}-sheet-${sheetId}`;
+}
+
+async function upsertSheetInDb(connector: ConnectorResource, sheet: Sheet) {
+  await GoogleDriveSheet.upsert({
+    connectorId: connector.id,
+    driveFileId: sheet.spreadsheet.id,
+    driveSheetId: sheet.id,
+    name: sheet.title,
+  });
+}
+
+async function upsertTable(
+  connector: ConnectorResource,
+  sheet: Sheet,
+  rows: string[][]
+) {
+  const dataSourceConfig = await dataSourceConfigFromConnector(connector);
+
+  const { id, spreadsheet, title } = sheet;
+  // Table name will be slugify in front.
+  const tableName = `${spreadsheet.title} - ${title}`;
+  const tableDescription = `Structured data from the Google Spreadsheet (${spreadsheet.title}) and sheet (${title}`;
+
+  const csv = stringify(rows);
+
+  // Upserting is safe: Core truncates any previous table with the same Id before
+  // the operation. Note: Renaming a sheet in Google Drive retains its original Id.
+  await upsertTableFromCsv({
+    dataSourceConfig,
+    tableId: makeTableIdFromSheetId(spreadsheet.id, id),
+    tableName,
+    tableDescription,
+    tableCsv: csv,
+    loggerArgs: {
+      connectorId: connector.id,
+      sheetId: id,
+      spreadsheetId: spreadsheet.id,
+    },
+  });
+}
+
+function makeRandomString(length: number) {
+  return uuidv4.toString().substring(0, length);
+}
+
+function findDataRangeAndSelectRows(allRows: string[][]): string[][] {
+  // Find the first row with data to determine the range.
+  const firstNonEmptyRow = allRows.find((row) =>
+    row.some((cell) => cell.trim() !== "")
+  );
+  if (!firstNonEmptyRow) {
+    return []; // No data found.
+  }
+
+  // Identify the range of data: Start at the first non-empty cell and end at the nearest following empty cell or row end.
+  const startIndex = firstNonEmptyRow.findIndex((cell) => cell.trim() !== "");
+  let endIndex = firstNonEmptyRow.findIndex(
+    (cell, idx) => idx > startIndex && cell.trim() === ""
+  );
+  if (endIndex === -1) {
+    endIndex = firstNonEmptyRow.length;
+  }
+
+  // Select only rows and columns within the data range.
+  return allRows
+    .map((row) => row.slice(startIndex, endIndex))
+    .filter((row) => row.some((cell) => cell.trim() !== ""));
+}
+
+function getSanitizedHeaders(rawHeaders: string[], loggerArgs: object) {
+  return rawHeaders.reduce<string[]>((acc, curr) => {
+    const slugifiedName = slugify(curr);
+
+    if (!acc.includes(slugifiedName)) {
+      acc.push(slugifiedName);
+    } else {
+      logger.info(
+        loggerArgs,
+        "Duplicated headers detected; suffixes added for uniqueness."
+      );
+
+      // Append a 4-character suffix to duplicate header names for uniqueness.
+      const randomSuffix = makeRandomString(4);
```

Comment:
technically we _could_ be super unlucky 😛

Reply:
Yep, and activity will be retried so we're good

Reply:
👌